docs/source-pytorch/fabric/advanced/collectives.rst

```diff
@@ -4,45 +4,237 @@
+   :width: 100%
+
+A barrier is needed when processes do different amounts of work and as a result fall out of sync.
+
+.. code-block:: python
+
+    fabric = Fabric(accelerator="cpu", devices=4)
+    fabric.launch()
+
+    # Simulate each process taking a different amount of time
+    sleep(2 * fabric.global_rank)
+    print(f"Process {fabric.global_rank} is done.")
+
+    # Wait for all processes to reach the barrier
     fabric.barrier()
+    print("All processes reached the barrier!")
+
+
+A more realistic scenario is when downloading data.
+Here, we need to ensure that processes only start to load the data once it has completed downloading.
+Since downloading should be done on rank 0 only to :ref:`avoid race conditions <race conditions>`, we need a barrier:
+
+.. code-block:: python
+
+    if fabric.global_rank == 0:
+        print("Downloading dataset. This can take a while ...")
+        download_dataset()
+
+    # All other processes wait here until rank 0 is done with downloading:
+    fabric.barrier()
+
+    # After everyone reached the barrier, they can access the downloaded files:
+    load_dataset()
+
+
+----
+
+.. _broadcast collective:
+
+*********
+Broadcast
+*********
+
+.. figure:: https://pl-flash-data.s3.amazonaws.com/fabric/docs/collectives/broadcast.jpeg
+   :alt: The broadcast collective operation
+   :width: 100%
 
+The broadcast operation sends a tensor of data from one process to all other processes so that all end up with the same data.
+
+.. code-block:: python
+
+    fabric = Fabric(...)
+
+    # Transfer a tensor from one process to all the others
+    result = fabric.broadcast(tensor)
+
+    # By default, the source is the process rank 0 ...
+    result = fabric.broadcast(tensor, src=0)
+
+    # ... which can be change to a different rank
+    result = fabric.broadcast(tensor, src=3)
+
+
+A concrete example:
+
+.. code-block:: python
+
+    fabric = Fabric(devices=4, accelerator="cpu")
+    fabric.launch()
+
+    # Data is different on each process
+    learning_rate = torch.rand(1)
+    print("Before broadcast:", learning_rate)
+
+    # Transfer the tensor from one process to all the others
+    learning_rate = fabric.broadcast(learning_rate)
+    print("After broadcast:", learning_rate)
+
+
+----
+
+
+******
+Gather
+******
+
+.. figure:: https://pl-flash-data.s3.amazonaws.com/fabric/docs/collectives/all-gather.jpeg
+   :alt: The All-gather collective operation
+   :width: 100%
+
+The gather operation transfers the tensors from each process to every other process and stacks the results.
+As opposed to the :ref:`broadcast <broadcast collective>`, every process gets the data from every other process, not just from a particular rank.
+
+.. code-block:: python
+
+    fabric = Fabric(...)
+
+    # Gather the data from
+    result = fabric.all_gather(tensor)
+
+    # By default, you can't back-propagate through the all-gather operation
+    result = fabric.all_gather(tensor, sync_grads=False)
```

Comment:
I wouldn't mention sync_grads here, as I believe we will remove it eventually in favor of using the `no_grad` context manager outside of the `all_gather`

Reply:
But that would be weird. I don't think torch would ever invert this behavior. It is like the reverse of what you want most of the time, this would be asking you to add `torch.no_grad()` most of the time, eventually becoming boilerplate code. I'm not sure what this idea is motivated by. Any pointers?

Reply:
We want to align our collectives API with PyTorch's and this is a difference that can become problematic. This boilerplate is managed by the PL loops, and in Fabric you will be able to easily turn on or off these context managers.

I don't mean to discuss this here. My suggestion is to not add docs that mention things that will likely go away